File: MATLAB/process_fastqs.m
function process_fastqs(input_fastq, outdir, tenX_version)
%process_fastqs splits Illumina sequencing runs of bacterial scRNAseq into
%a format recognized by CellRanger using both 10X and Probe UMIs
%
%   process_fastqs(INPUT_FASTQ, OUTDIR, TENX_VERSION) splits INPUT_FASTQ
%   into R1 and R2 reads appropriate for CellRanger and places the resulting 
%   FastQ files in OUTDIR/10X (10X UMIs in R1) and OUTDIR/Probe
%   (Probe UMIs in R1). Specify the TENX_VERSION used to 
%   prepare the library (2 or 3), to set the UMI length trimmed from the
%   INPUT_FASTQ.
%
%   Author: Duluxan Sritharan. Hormoz Lab. Harvard Medical School.

    tic;
    
    assert(exist(input_fastq, 'file')==2);
    
    assert(~(exist(outdir, 'dir')==7), 'Output path already exists');
    mkdir(outdir);
    mkdir(sprintf('%s/10X', outdir));
    mkdir(sprintf('%s/Probe', outdir));    
    
    fprintf('Processing: %s\n', input_fastq);
    
    Extender = seqrcomplement('CATAGTTTCTTCGAGCAA');
    L_ext = length(Extender);
    L_CB = 16;    
    if (tenX_version==2)
        L_UMI = 10;
    elseif (tenX_version==3)
        L_UMI = 12;
    else
        error('Unrecognized 10X version');
    end
    
    [input_fastq, ext] = maybe_unzip(input_fastq);
    info = fastqinfo(input_fastq);
    N_reads = double(info.NumberOfEntries);
    block_size = 1e6;
    N_blocks = ceil(N_reads/block_size);
    fprintf('Splitting %10d reads over %4d blocks\n', N_reads, N_blocks);
    
    for b = 1:N_blocks
        tic;        
        lo = (b-1)*block_size+1;
        hi = min(b*block_size, N_reads);        
        [L_orig{b}, masks{b}, probe_umi_end_pos{b}, sc{b}] = ...
            process_per_block(input_fastq, lo, hi, Extender, L_CB, L_UMI, L_ext, outdir);
        fprintf('Block: %4d/%4d. Reads: [%10d, %10d] in %6d seconds\n', b, N_blocks, lo, hi, toc);
    end   

    L_orig = vertcat(L_orig{:});
    masks = vertcat(masks{:});
    probe_umi_end_pos = vertcat(probe_umi_end_pos{:});
    sc = vertcat(sc{:});
    
    maybe_clear_unzipped(input_fastq, ext);
    
    fprintf('Output R2...\n');    
    
    output_fastq2 = sprintf('%s/10X/R2_001.fastq', outdir);
    info = fastqinfo(output_fastq2);
    assert(double(info.NumberOfEntries) ==  length(vertcat(masks.valid_seq)));
    gzip(output_fastq2);
    delete(output_fastq2);
    copyfile(sprintf('%s.gz', output_fastq2), strrep(sprintf('%s.gz', output_fastq2), '10X', 'Probe'));
    
    fprintf('Output 10X R1...\n');    
    
    output_fastq1 = sprintf('%s/10X/R1_001.fastq', outdir);
    info = fastqinfo(output_fastq1);
    assert(double(info.NumberOfEntries) == length(vertcat(masks.valid_seq)));    
    gzip(output_fastq1);
    delete(output_fastq1);
    
    fprintf('Output Probe R1...\n');    
    
    output_fastq1 = sprintf('%s/Probe/R1_001.fastq', outdir);
    info = fastqinfo(output_fastq1);
    assert(double(info.NumberOfEntries) == length(vertcat(masks.valid_seq)));
    gzip(output_fastq1);
    delete(output_fastq1);
    
    fprintf('Output provenance...\n');    
       
    metric_file = sprintf('%s/meta.mat', outdir);
    save(metric_file, 'input_fastq', 'L_orig', 'L_CB', 'L_UMI', 'L_ext', 'masks', 'sc', 'probe_umi_end_pos', '-v7.3', '-nocompression');
    
    fprintf('%f seconds\n', toc);
    
end

function [L_orig, masks, probe_umi_end_pos, sc] = ...
    process_per_block(input_fastq, lo, hi, Extender, L_CB, L_UMI, L_ext, outdir)
    
    fprintf('\tReading S...\n');
    [~, S, ~] = fastqread(input_fastq, 'Blockread', [lo, hi]);
    S=S';
    
    fprintf('\tTrimming by length...\n');    
    L_orig = cellfun(@length, S);
    masks.full_seq = L_orig > 110;    
    S = S(masks.full_seq);    
    masks.full_seq = find(masks.full_seq);
    
    fprintf('\tAligning...\n');    
    [sc, al] = nwalign(S, Extender, 'Alphabet', 'NT', 'glocal', true);
    masks.extender_found = (sc>=17);
    al = al(masks.extender_found);    
    S = S(masks.extender_found);  
    masks.extender_found = masks.full_seq(masks.extender_found);
    
    fprintf('\tTrimming by extender...\n');    
    probe_umi_end_pos = cellfun(@(x) sum(x(1,1:find(x(3,:) ~= '-', 1, 'first')-1)~='-'), al);
    clear al;
    % Should be a polyT region separating separating 10X (CB/UMI) and Probe UMI
    probe_mask = find(probe_umi_end_pos >= L_CB+2*L_UMI);
    S = S(probe_mask);    
    masks.valid_probe = masks.extender_found(probe_mask);
    
    % Keep the extender since the sequences are ~30 bps without it which
    % may affect alignment, but remove sequences consisting of just the
    % extender
    seq_mask = (L_orig(masks.valid_probe)>probe_umi_end_pos(probe_mask)+L_ext);
    S = S(seq_mask);
    masks.valid_seq = masks.valid_probe(seq_mask);
    
    fprintf('\tSplitting S...\n');
    
    S_CB  = cellfun(@(x)   x(1:L_CB), S, 'un', false);
    S_10X_UMI = cellfun(@(x) x(L_CB+1:L_CB+L_UMI), S, 'un', false);
    S_probe_UMI = cellfun(@(x,e) x(e-L_UMI+1:e), S, num2cell(probe_umi_end_pos(probe_mask(seq_mask))), 'un', false);
    S = cellfun(@(x,s) x(s+1:end), S, num2cell(probe_umi_end_pos(probe_mask(seq_mask))), 'un', false);
    
    fprintf('\tReading Q...\n');
    [~, ~, Q] = fastqread(input_fastq, 'Blockread', [lo, hi]);
    Q=Q(masks.valid_seq)';
    
    fprintf('\tSplitting Q...\n');
    
    Q_CB  = cellfun(@(x)   x(1:L_CB), Q, 'un', false);
    Q_10X_UMI = cellfun(@(x) x(L_CB+1:L_CB+L_UMI), Q, 'un', false);
    Q_probe_UMI = cellfun(@(x,e) x(e-L_UMI+1:e), Q, num2cell(probe_umi_end_pos(probe_mask(seq_mask))), 'un', false);
    Q = cellfun(@(x,s) x(s+1:end), Q, num2cell(probe_umi_end_pos(probe_mask(seq_mask))), 'un', false);
    
    fprintf('\tReading H...\n');
    [H, ~, ~] = fastqread(input_fastq, 'Blockread', [lo, hi]);
    H=H(masks.valid_seq)';
    
    fprintf('\tOutput R2...\n');    
    
    output_fastq2 = sprintf('%s/10X/R2_001.fastq', outdir);           
    fastqwrite(output_fastq2, H, S, Q);    
    
    fprintf('\tOutput 10X R1...\n');    
    
    output_fastq1 = sprintf('%s/10X/R1_001.fastq', outdir);        
    S = strcat(S_CB, S_10X_UMI);
    Q = strcat(Q_CB, Q_10X_UMI);
    fastqwrite(output_fastq1, H, S, Q);
        
    fprintf('\tOutput Probe R1...\n');    
    
    output_fastq1 = sprintf('%s/Probe/R1_001.fastq', outdir);
    S = strcat(S_CB, S_probe_UMI);
    Q = strcat(Q_CB, Q_probe_UMI);
    fastqwrite(output_fastq1, H, S, Q);    
    
end

File: Python/process_fastqs.py
import os
import gzip
import shutil
import time
import pickle
from subprocess import check_output
from datetime import timedelta

import numpy as np

import parasail

from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq

from utils import *


def process_fastqs(input_fastq, outdir, tenX_version):

    start_time = time.time()

    assert os.path.isfile(input_fastq), 'Invalid path to 10x counts file'

    assert not os.path.isdir(outdir), 'Output path already exists'

    os.makedirs(outdir)
    os.makedirs(f'{outdir}/10X')
    os.makedirs(f'{outdir}/Probe')
    os.makedirs('meta')

    print(f'Processing {input_fastq}')

    complement = {
        'A': 'T',
        'C': 'G',
        'G': 'C',
        'T': 'A',
    }
    Extender = ''.join(
        [complement.get(nuc) for nuc in reversed('CATAGTTTCTTCGAGCAA')]
    )
    L_ext = len(Extender)
    L_CB = 16
    if tenX_version == 2:
        L_UMI = 10
    elif tenX_version == 3:
        L_UMI = 12
    else:
        raise ValueError('Unrecognized 10X version')

    if input_fastq.endswith('gz'):
        print('Unzipping FASTQ file')
        with gzip.open(input_fastq, 'rt') as f_in:
            with open(input_fastq[:-3], 'wt') as f_out:
                shutil.copyfileobj(f_in, f_out)
        input_fastq = input_fastq[:-3]

    command = f'wc -l {input_fastq}'
    N_reads = int(int(check_output(command, shell=True).split()[0])/4)
    block_size = 1e7
    N_blocks = int(np.ceil(N_reads/block_size))
    print(f'Splitting {N_reads} over {N_blocks}')

    record_iter = SeqIO.parse(input_fastq, 'fastq')

    L_orig = []
    masks = {
        'full_seq': [],
        'extender_found': [],
        'valid_probe': [],
        'valid_seq': [],
    }
    probe_umi_end_pos = []
    sc = []

    for records in batch_iterator(record_iter, block_size):
        _L, _ma, _pr, _sc = \
            process_per_block(records, Extender, L_CB, L_UMI, L_ext, outdir)
        L_orig.append(_L)
        for k in masks.keys():
            masks[k].append(_ma[k])
        probe_umi_end_pos.append(_pr)
        sc.append(_sc)

    L_orig = np.hstack(L_orig)
    for k in masks.keys():
        masks[k] = np.hstack(masks[k])
    probe_umi_end_pos = np.hstack(probe_umi_end_pos)
    sc = np.hstack(probe_umi_end_pos)

    command = f'wc -l {outdir}/10X/R2_001.fastq'
    N_reads_processed = int(
        int(check_output(command, shell=True).split()[0])/4)

    print('Removing unzipped FASTQ file')
    os.remove(input_fastq)

    assert len(masks['valid_seq']) == N_reads_processed

    print('Output and compress R2...')

    with open(f'{outdir}/10X/R2_001.fastq', 'rt') as f_in:
        with gzip.open(f'{outdir}/Probe/R2_001.fastq.gz', 'wt') as f_out:
            shutil.copyfileobj(f_in, f_out)
    with open(f'{outdir}/10X/R2_001.fastq', 'rt') as f_in:
        with gzip.open(f'{outdir}/10X/R2_001.fastq.gz', 'wt') as f_out:
            shutil.copyfileobj(f_in, f_out)
    os.remove(f'{outdir}/10X/R2_001.fastq')

    print('Compress 10X R1...')

    with open(f'{outdir}/10X/R1_001.fastq', 'rt') as f_in:
        with gzip.open(f'{outdir}/10X/R1_001.fastq.gz', 'wt') as f_out:
            shutil.copyfileobj(f_in, f_out)
    os.remove(f'{outdir}/10X/R1_001.fastq')

    print('Compress Probe R1...')

    with open(f'{outdir}/Probe/R1_001.fastq', 'rt') as f_in:
        with gzip.open(f'{outdir}/Probe/R1_001.fastq.gz', 'wt') as f_out:
            shutil.copyfileobj(f_in, f_out)
    os.remove(f'{outdir}/Probe/R1_001.fastq')

    print('Output provenance...')

    metric = {
        'L_orig': L_orig,
        'masks': masks,
        'probe_umi_end_pos': probe_umi_end_pos,
        'sc': sc,
    }

    pickle.dump(metric, open('meta/meta.pkl', 'wb'))

    time_elapsed = time.time() - start_time
    print(f'FASTQ processed in: {timedelta(seconds=time_elapsed)}')


# def batch_iterator(iterator, batch_size):
#    entry = True  # Make sure we loop once
#    while entry:
#        batch = []
#        while len(batch) < batch_size:
#            try:
#                entry = next(iterator)
#            except StopIteration:
#                entry = None
#            if entry is None:
#                # End of file
#                break
#            batch.append(entry)
#        if batch:
#            yield batch


def process_per_block(records, Extender, L_CB, L_UMI, L_ext, outdir):
    S = np.array([rec.seq for rec in records], dtype=object)

    print('Trimming by length...')
    L_orig = np.array([len(seq) for seq in S])
    full_seq = np.where(L_orig > 110)[0]
    S = S[full_seq]

    print('Aligning...')
    sc, al = [], []
    for s in S:
        result = parasail.sg_trace(s._data, Extender, 8, 8, parasail.nuc44)
        sc.append(0.277316 * result.score)
        al.append(result.cigar.decode.split(b'I')[0])
    extender_found = np.where(np.array(sc) >= 17)[0]
    al = np.array(al)[extender_found]
    S = S[extender_found]
    extender_found = full_seq[extender_found]

    print('Trimming by extender...')
    probe_umi_end_pos = np.array([int(match) for match in al])
    # Should be a polyT region separating 10X (CB/UMI) and Probe UMI
    probe_mask = np.where(probe_umi_end_pos >= L_CB + 2*L_UMI)[0]
    S = S[probe_mask]
    valid_probe = extender_found[probe_mask]

    # Keep the extender since the sequences are ~30 bps without it which
    # may affect alignment, but remove sequences consisting of just the
    # extender
    seq_mask = np.where(L_orig[valid_probe] >
                        probe_umi_end_pos[probe_mask] + L_ext)[0]
    S = S[seq_mask]
    valid_seq = valid_probe[seq_mask]

    print('Splitting S...')
    S_CB = np.array([seq[:L_CB]._data for seq in S])
    S_10X_UMI = np.array([seq[L_CB:L_CB+L_UMI]._data for seq in S])
    S_probe_UMI = np.array(
        [seq[e-L_UMI:e]._data
         for (seq, e) in zip(S, probe_umi_end_pos[probe_mask[seq_mask]])]
    )
    S = [seq[s:]._data
         for (seq, s) in zip(S, probe_umi_end_pos[probe_mask[seq_mask]])]

    print('Reading Q...')
    Q = np.array(
        [rec.letter_annotations['phred_quality'] for rec in records],
        dtype=object,
    )
    Q = Q[valid_seq]

    print('Splitting Q...')
    Q_CB = np.array([x[:L_CB] for x in Q])
    Q_10X_UMI = np.array([x[L_CB:L_CB+L_UMI] for x in Q])
    Q_probe_UMI = np.array(
        [x[e-L_UMI:e]
         for (x, e) in zip(Q, probe_umi_end_pos[probe_mask[seq_mask]])]
    )
    Q = [x[s:] for (x, s) in zip(Q, probe_umi_end_pos[probe_mask[seq_mask]])]

    print('Reading H...')
    H = np.array([rec.description for rec in records])
    H = H[valid_seq]

    print('Output R2...')

    output_fastq2 = []
    for (s, h, q) in zip(S, H, Q):
        record = SeqRecord(Seq(s), id=h, name=h, description='')
        record.letter_annotations['phred_quality'] = q
        output_fastq2.append(record)
    with open(f'{outdir}/10X/R2_001.fastq', 'at+') as f:
        SeqIO.write(output_fastq2, f, 'fastq')

    print('Output 10X R1...')
    S = np.char.add(S_CB, S_10X_UMI)
    Q = np.hstack((Q_CB, Q_10X_UMI))
    output_fastq1 = []
    for (s, h, q) in zip(S, H, Q):
        record = SeqRecord(Seq(s), id=h, name=h, description='')
        record.letter_annotations['phred_quality'] = q
        output_fastq1.append(record)
    with open(f'{outdir}/10X/R1_001.fastq', 'at+') as f:
        SeqIO.write(output_fastq1, f, 'fastq')

    print('Output Probe R1...')
    S = np.char.add(S_CB, S_probe_UMI)
    Q = np.hstack((Q_CB, Q_probe_UMI))
    output_fastq1 = []
    for (s, h, q) in zip(S, H, Q):
        record = SeqRecord(Seq(s), id=h, name=h, description='')
        record.letter_annotations['phred_quality'] = q
        output_fastq1.append(record)
    with open(f'{outdir}/Probe/R1_001.fastq', 'at+') as f:
        SeqIO.write(output_fastq1, f, 'fastq')

    masks = {
        'full_seq': full_seq,
        'extender_found': extender_found,
        'valid_probe': valid_probe,
        'valid_seq': valid_seq,
    }

    return L_orig, masks, probe_umi_end_pos, sc
